### PR TITLE
PQR Parser

### PR DESCRIPTION
**Pull request by zisikons**
This pull request addresses issue #2226, introducing support for parsing files in the [PQR format](https://sites.google.com/a/poissonboltzmann.org/software/file-formats/biomolecular-structurw/pqr).

We look forward to your feedback :).

- [X] I hereby agree to dual licence this and any previous contributions under both
the _Biopython License Agreement_ **AND** the _BSD 3-Clause License_.

- [X] I have read the ``CONTRIBUTING.rst`` file, have run ``flake8`` locally, and
understand that AppVeyor and TravisCI will be used to confirm the Biopython unit
tests and style checks pass with these changes.

- [X] I have added my name to the alphabetical contributors listings in the files
``NEWS.rst`` and ``CONTRIB.rst`` as part of this pull request, am listed
already, or do not wish to be listed. (*This acknowledgement is optional.*)

### CODE REVIEW

**Comment by peterjc** [comment]
Looks like some (easily fixed) style checks:

```
$ flake8
./Bio/PDB/PDBIO.py:241:29: W291 trailing whitespace
./Bio/PDB/PDBIO.py:242:29: W291 trailing whitespace
./Bio/PDB/PDBIO.py:243:22: W291 trailing whitespace
./Bio/PDB/PDBIO.py:244:24: W291 trailing whitespace
./Bio/PDB/PDBIO.py:245:25: W291 trailing whitespace
./Bio/PDB/PDBIO.py:247:24: W291 trailing whitespace
./Bio/PDB/PDBIO.py:248:23: W291 trailing whitespace
./Bio/PDB/PDBIO.py:249:19: W291 trailing whitespace
./Bio/PDB/PDBIO.py:250:19: W291 trailing whitespace
./Bio/PDB/PDBIO.py:251:19: W291 trailing whitespace
./Bio/PDB/PDBIO.py:252:24: W291 trailing whitespace
./Bio/PDB/PDBIO.py:253:24: W291 trailing whitespace
./Bio/PDB/PQRParser.py:42:45: Q000 Remove bad quotes
```

Some editors can remove trailing spaces for you.

File: Bio/PDB/PQRParser.py [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+        structure_builder = self.structure_builder
+        current_model_id = 0
+        # Flag we have an open model
+        model_open = 0
+        current_chain_id = None
+        current_segid = None
+        current_residue_id = None
+        current_resname = None
+        for i in range(0, len(coords_trailer)):
+            line = coords_trailer[i].rstrip('\n')
```
Comment (peterjc) [inline Bio/PDB/PQRParser.py]: We prefer double quotes over single quotes (something the code formatting tool ``black`` does, which we are working towards adopting for the whole of Biopython).

**Comment by peterjc** [comment]
@jgreener64 can you take a look please?

**Comment by jgreener64** [comment]
Sure, I'll have a look within the next few days.

**Comment by zisikons** [comment]
I resolved the linting issues and now all the tests are successful. Apparently, in the previous version `PDBIO.py` was not being linted inside a test script I was using.

**Comment by jgreener64** [comment]
Overall this looks good and I think is a nice addition. I like the changes to `Atom.py` and `PDBIO.py` because they extend the existing methods, and the tests look good.

I see what you have done with sub-classing `PDBParser` and rewriting the `parse_coordinates` method. However that function is very large and so there is quite a lot of code duplication, which could cause problems if someone later updates one `parse_coordinates` method but forgets the other. Is there any way to split up this function so the over-ridden method for `PQRParser` is shorter? You could pull out the different bit into its own function.

Similarly I wonder if `_init_atom` and `_init_atom_from_pqr` can be combined, for example by having a keyword `pqr` argument to the `_init_atom` function that switches behaviour as needed.

A couple of smaller things:
- If you write a PQR file with a `None` radius, e.g. you accidentally try and write a PDB object, does it warn sensibly?
- The `charge` property of an `Atom` might appear to many people to be the value in columns 79-80 of the PDB file, which the Biopython parser does not read. In fact it is something else entirely, which should be made clear in the documentation of `Atom`. I would even consider storing the variable internally as `pqr_charge` to be clear that this is not the PDB charge.
- Could you add a mention in the docs that you can now parse PQR files, with an example?

I think this is a good PR I'm just careful to add duplicated lines to the project, as they have caught out maintainers later on before.

**Comment by artbendandi** [comment]
As suggested, we have implemented a flag-based functionality on the PDB Parser, in order to avoid duplicate code.
We have introduced a Warning for invalid or missing radius and charge values.
We renamed charge to pqr_charge when refered to the charge of the PQR file, to avoid misunderstandings.
We added a section in the docs on how to parse and how to write PQR files.

We thank you for your positive comments, encouragement and suggestions.

**Comment by jgreener64** [comment]
Thanks for making these changes, this now looks ready to me.

**Comment by jgreener64** [comment]
Although you could also add a line to NEWS.md outlining these changes, and add your names there.

**Comment by artbendandi** [comment]
We updated NEWS.rst accordingly, and resolved the resulting issues.

**Comment by jgreener64** [comment]
Thanks, this now looks good to me @peterjc.

**Comment by artbendandi** [comment]
Thank you :)

**Comment by peterjc** [comment]
@artbendandi do you also agree to the dual licensing?

I could do a squash-and-merge, but that will discard the separate authorship here (and the commit history), so I would prefer not to here.

I can't do a rebase-and-merge directly from GitHub due to the merge conflict in the NEWS file. I can fix this easily at the command line with ``git rebase origin/master`` but can't push to your branch (there is an option in GitHub when you submit a pull request to allow edits from the maintainers).

Are either of you comfortable with ``git rebase``?

Alternatively, I could do the rebase directly to master, and just close the PR afterwards?

**Comment by artbendandi** [comment]
@peterjc I assume that by dual licensing you mean that the contribution to Biopython would be accredited to both @zisikons and me. To avoid any misunderstanding however, could you clarify what you mean by that?

We are fine with doing a git rebase ourselves, but the option you mention is enabled on our pull request, so you should be able to push to our branch (I think). 

Thank you for considering the dual contribution of @zisikons and myself in this pull request. Indeed, it would be a pity if the joint authorship were to be lost.

**Comment by peterjc** [comment]
We need both of you to agree to dual license your contributions, @zisikons did in the first pull request message - @artbendandi has not yet.

If either of you are confident with ``git rebase``, please go ahead and rebase onto the main master branch. Thanks!

**Comment by artbendandi** [comment]
- [X]  I hereby agree to dual licence this and any previous contributions under both
the Biopython License Agreement AND the BSD 3-Clause License.

- [X]  I have read the CONTRIBUTING.rst file, have run flake8 locally, and
understand that AppVeyor and TravisCI will be used to confirm the Biopython unit
tests and style checks pass with these changes.

- [X]  I have added my name to the alphabetical contributors listings in the files
NEWS.rst and CONTRIB.rst as part of this pull request, am listed
already, or do not wish to be listed. (This acknowledgement is optional.)

**Comment by artbendandi** [comment]
A couple of days ago we performed git rebase. Naturally, we remain available if anything further is needed. We thank you for your comments and input during this pull request.

**Comment by artbendandi** [comment]
@peterjc We wanted to write a quick message because we noticed that the PR is still "Checking for ability to merge automatically...", and therefore is still open. We were wondering if we misunderstood, and should've taken further actions besides the force-push. Should we do the merge ourselves? 
Thank you and apologies for the mixup.

**Comment by peterjc** [comment]
Looking over the history both @jgreener64 and @JoaoRodrigues approved this so it looks good to go other than GitHub showing a conflict in NEWS.rst

If you can do another git rebase, that'd be ideal - then any one of the core team can push the big green "rebase-and-merge" button and keep the commit history (with both contributors clearly attributed).

**Comment by peterjc** [comment]
P.S. Sorry that this got lost over Christmas, we should have merged it a month ago.

**Comment by artbendandi** [comment]
@peterjc We resolved the conflict in the NEWS.rst but now the checks are failing and we have no idea why, it should've been fine since we did not change anything else. Can you check into that?

**Comment by peterjc** [comment]
The flake8 error isn't your fault, looks like a plugin issue:
https://github.com/PyCQA/flake8-bugbear/issues/108

**Comment by artbendandi** [comment]
Ok, so how should we proceed?

**Comment by peterjc** [comment]
Happily flake8-bugbear was updated overnight, re-running TravisCI

The second problem is due to changes on the master branch with the goal of removing ``Bio._py3k`` now that we've dropped Python 2 support:

```
test_PQR ... loading tests failed:
Failed to import test module: test_PQR
Traceback (most recent call last):
  File "C:\Py\lib\unittest\loader.py", line 154, in loadTestsFromName
    module = __import__(module_name)
  File "C:\projects\biopython\Tests\test_PQR.py", line 15, in <module>
    from Bio._py3k import StringIO
ImportError: cannot import name 'StringIO' from 'Bio._py3k' (C:\projects\biopython\build\lib.win32-3.7\Bio\_py3k\__init__.py)
```

You can now just use ``from io import StringIO`` instead.

**Comment by artbendandi** [comment]
@peterjc looks like we're good to go :)

**Comment by jgreener64** [comment]
If this is ready to go could someone with merge rights merge it please?

**Pull request merged**